### PR TITLE
Adds push build pipelines to dronegen

### PR DESCRIPTION
**Pull request by webvictim**
@awly: Just raising this against your branch to get a review on the general code style and whether you think this is the right way to go about this. The `.drone.yml` diff looks OK to me so don't worry too much about reviewing that part.

Even If I could migrate all the `push-build-*` and `build-*` pipelines to `dronegen` it'd make my job a **ton** easier.

This is pretty addictive...

### CODE REVIEW

File: dronegen/tests.go [hunk 1]
```diff
@@ -54,8 +54,7 @@ func testCodePipeline() pipeline {
 				`mkdir -p /tmpfs/go/src/github.com/gravitational/teleport /tmpfs/go/cache`,
 				`cd /tmpfs/go/src/github.com/gravitational/teleport`,
 				`git init && git remote add origin ${DRONE_REMOTE_URL}`,
-				`
-# handle pull requests
+				`# handle pull requests
```
Comment (webvictim) [inline dronegen/tests.go:57]: These changes (removing the initial newline) are just to make the formatting in `.drone.yml` a bit nicer as VS Code's YAML syntax highlighting seems to break when there's a newline in a YAML multiline `- |2` block.

It basically changes this:

```yaml
- |2

  git checkout...
```

to this:
```yaml
- |
  git checkout
```

**Comment by webvictim** [comment]
I'm also pondering whether there's a way to make the unmarshal functions indent the YAML by two spaces (as in the current Drone pipeline) to make reading the diffs easier due to line juxtaposition. I guess the question is whether it's easier to modify `dronegen` or whether I just go and re-indent the existing `.drone.yml`.

File: dronegen/push.go [hunk 2]
```diff
@@ -1,6 +1,175 @@
...
+	} else {
+		if params.arch == "arm" {
+			makefileTarget = "release-arm"
+			extraDockerImage = "quay.io/gravitational/teleport-buildbox-arm:$RUNTIME"
+		} else if params.arch == "arm64" {
+			makefileTarget = "release-arm64"
+			extraDockerImage = "quay.io/gravitational/teleport-buildbox-arm:$RUNTIME"
+		}
+		makeCommand = fmt.Sprintf(`make -C build.assets %s OS=$OS ARCH=$ARCH`, makefileTarget)
+	}
```
Comment (webvictim) [inline dronegen/push.go:58]: I don't like this section much, I originally had a `switch (params.arch)` block but the FIPS part complicates it somewhat.

File: dronegen/push.go [hunk 3]
```diff
@@ -1,6 +1,175 @@
...
+
+	commands = append(commands, fmt.Sprintf(`docker pull %s || true`, baseDockerImage))
+	if extraDockerImage != "" {
+		commands = append(commands, fmt.Sprintf(`docker pull %s || true`, extraDockerImage))
+	}
+	commands = append(commands, `cd /go/src/github.com/gravitational/teleport`)
+	commands = append(commands, makeCommand)
+
+	return commands
+}
```
Comment (webvictim) [inline dronegen/push.go:68]: This logic is likely to be reused for tag build pipelines too so can be moved elsewhere.

File: dronegen/push.go [hunk 4]
```diff
@@ -1,6 +1,175 @@
...
+	return []pipeline{
+		pushPipeline(buildType{os: "linux", arch: "amd64"}),
+		pushPipeline(buildType{os: "linux", arch: "i386"}),
+		pushPipeline(buildType{os: "linux", arch: "amd64", fips: true}),
+		pushPipeline(buildType{os: "windows", arch: "amd64"}),
+		pushPipeline(buildType{os: "linux", arch: "arm"}),
+		pushPipeline(buildType{os: "linux", arch: "arm", fips: true}),
+		pushPipeline(buildType{os: "linux", arch: "arm64"}),
+		pushPipeline(buildType{os: "linux", arch: "arm64", fips: true}),
+	}
```
Comment (awly) [inline dronegen/push.go:15]:
```suggestion
	var ps []pipeline
	for _, arch := range []string{"amd64", "386", "arm", "arm64"} {
		for _, fips := []bool{false, true} {
			if arch == "386" && fips {
				// FIPS mode not supported on i386
				continue
			}
			ps = append(ps, pushPipeline(buildType{os: "linux", arch: arch, fips: fips})
		}
	}
	// Only amd64 Windows is supported for now.
	ps = append(ps, pushPipeline(buildType{os: "windows", arch: "amd64"}))
	return ps
```

File: dronegen/push.go (hunk 2 above)
Comment (awly) [reply dronegen/push.go:58]: How about we add alias make rules to make target naming consistent?
Add something like:
```
release-amd64: release
release-386: release
release-amd64-fips: release-fips
release-386-fips: release-fips
```
(maybe also force the `ARCH` value)

Then you could do:
```
makefileTarget := fmt.Sprintf("release-%s", params.arch)
if params.fips {
  makefileTarget += "-fips"
}
```

And something similar for docker images.

File: dronegen/push.go [hunk 5]
```diff
@@ -1,6 +1,175 @@
...
+		return "386"
+	}
+	return arch
+}
+
+func pushBuildCommands(params buildType) []string {
+	commands := []string{
+		`apk add --no-cache make`,
+		`chown -R $UID:$GID /go`,
+	}
```
Comment (awly) [inline dronegen/push.go:32]: nit: move this further down, once `makeCommand` is set
that way you'll need less appending.

File: dronegen/push.go [hunk 6]
```diff
@@ -1,6 +1,175 @@
...
+		}
+		makeCommand = fmt.Sprintf(`make -C build.assets %s OS=$OS ARCH=$ARCH`, makefileTarget)
+	}
+
+	commands = append(commands, fmt.Sprintf(`docker pull %s || true`, baseDockerImage))
+	if extraDockerImage != "" {
+		commands = append(commands, fmt.Sprintf(`docker pull %s || true`, extraDockerImage))
+	}
+	commands = append(commands, `cd /go/src/github.com/gravitational/teleport`)
+	commands = append(commands, makeCommand)
```
Comment (awly) [inline dronegen/push.go:65]:
```suggestion
	commands = append(commands, 
		`cd /go/src/github.com/gravitational/teleport`,
		makeCommand,
	)
```

File: dronegen/push.go [hunk 7]
```diff
@@ -1,6 +1,175 @@
...
+	pipelineName := fmt.Sprintf("push-build-%s-%s", params.os, params.arch)
+	pushEnvironment := map[string]value{
+		"UID":    value{raw: "1000"},
+		"GID":    value{raw: "1000"},
+		"GOPATH": value{raw: "/go"},
+		"OS":     value{raw: params.os},
+		"ARCH":   value{raw: translatePipelineArch(params.arch)},
+	}
+	if params.fips {
+		pipelineName = fmt.Sprintf("push-build-%s-%s-fips", params.os, params.arch)
```
Comment (awly) [inline dronegen/push.go:110]:
```suggestion
		pipelineName += "-fips"
```

**Review by awly — COMMENTED** [review]
Looks good overall.
One meta-comment: when code ends up with too many conditionals to handle inconsistent naming elsewhere, consider fixing that external naming

File: dronegen/push.go (hunk 2 above)
Comment (webvictim) [reply dronegen/push.go:58]: This makes sense.

**Comment by webvictim** [comment]
- consolidated the Makefile targets
- also took the opportunity to automatically pull buildbox images when running in Drone so that we don't need to handle that manually
  - it'll now be done for us when calling `make -C build.assets release-amd64-fips` or similar targets

File: dronegen/common.go [hunk 8]
```diff
@@ -15,3 +22,36 @@ var (
...
+func getMakefileTarget(params buildType) string {
+	makefileTarget := fmt.Sprintf("release-%s", params.arch)
+	if params.fips {
+		makefileTarget += "-fips"
+	}
+	return makefileTarget
+}
+
+// getBuildCheckoutCommands builds a list of commands for Drone to check out a git commit
+func getBuildCheckoutCommands(fips bool) []string {
```
Comment (awly) [inline dronegen/common.go:36]: nit: avoid a `get` prefix in function names

File: dronegen/common.go [hunk 9]
```diff
@@ -15,3 +22,36 @@ var (
...
 )
+
+// getMakefileTarget gets the correct Makefile target for a given arch/fips combo
+func getMakefileTarget(params buildType) string {
+	makefileTarget := fmt.Sprintf("release-%s", params.arch)
+	if params.fips {
+		makefileTarget += "-fips"
+	}
+	return makefileTarget
+}
```
Comment (awly) [inline dronegen/common.go:33]: this probably belongs in `push.go`

I intended this file to contain vars/functions re-used by multiple files (e.g. steps to checkout code, or common volumes like tmpfs and docker).
If a function is specific to a single set of pipelines (like `push-` ones), put it as close as possible to the caller

**Review by awly — APPROVED** [review]

File: dronegen/common.go (hunk 9 above)
Comment (webvictim) [reply dronegen/common.go:33]: My intention was to use this function in both `push.go` and `build.go` when I start migrating the next set of pipelines; I might be moving it a bit early.

**Pull request merged**